### PR TITLE
Implement configuration cache cleanup

### PR DESCRIPTION
**Pull request by eskatos**
in order not to fill up users disks.

This PR also moves the gradle version from the disk layout to the cache key to align the disk layout with other caches and let the cleanup implementation be simpler.

Cache entries created before this PR will be cleaned up too.

See individual commits for details.

### CODE REVIEW

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    fun CacheRepository.configurationCache() =
+        cache(cacheRootDir).withDisplayName("Configuration Cache")
+
+    private
+    fun CacheBuilder.withOnDemandLockMode() =
+        withLockOptions(LockOptionsBuilder.mode(FileLockManager.LockMode.OnDemand))
+
+    private
+    fun CacheBuilder.withLruCacheCleanup(): CacheBuilder =
+        withCleanup(cacheCleanupFactory.create(LeastRecentlyUsedCacheCleanup(
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt:72]: :x: I find it a bit hard to read this layout with all parenthesis in a single line. Can we use the usual nested layout here?

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                            CheckedFingerprint.Valid
+                        }
+                        else -> CheckedFingerprint.Invalid(invalidReason)
+                    }
+                }
+            }
+        })
+
+    fun useForStateLoad(cacheKey: String, action: (File) -> Unit) {
+        cache.withFileLock {
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt:111]: ⭕ The usages of `cache.withFileLock { ... cache.baseDirFor(cacheKey) ... }`  could be deduped with a `withBaseCacheDirFor(cacheKey) { ... }`

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    fun useForFingerprintCheck(cacheKey: String, check: (File) -> String?): CheckedFingerprint =
+        cache.withFileLock(Factory {
+            val fingerprint = cache.baseDirFor(cacheKey).fingerprintFile
+            when {
+                !fingerprint.isFile -> CheckedFingerprint.NotFound
+                else -> {
+                    when (val invalidReason = check(fingerprint)) {
+                        null -> {
+                            fileAccessTracker.markAccessed(fingerprint) // Prevent cleanup before state load
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt:101]: ❓ Shouldn't the `stateFile` be marked here as well? Is there any chance the cache cleanup action would be triggered in between?

Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt:101]: `SingleDepthFileAccessTracker.markAccessed()` will mark the parent directory. In other words, each cache entry directory (containing state & fingerprint) is a cleanup unit.

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt (hunk 1 above)
Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt]: Reads better?

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/DefaultInstantExecution.kt [hunk 4]
```diff
@@ -462,35 +465,8 @@ class DefaultInstantExecution internal constructor(
...
-    fun currentGradleVersion(): String =
-        GradleVersion.current().version
-
-    private
-    fun absoluteFile(path: String) =
-        File(rootDirectory, path).absoluteFile
-
-    // Skip instant execution for buildSrc for now. Should instead collect up the inputs of its tasks and treat as task graph cache inputs
+    // Skip instant execution for buildSrc for now.
+    // TODO Should instead collect up the inputs of its tasks and treat as task graph cache inputs
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/DefaultInstantExecution.kt]: Thanks, removed

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt (hunk 2 above)
Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt]: Done

**Comment by eskatos** [comment]
@bot-gradle test RFM

@bamboo PTAL

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt (hunk 3 above)
Comment (bamboo) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt:105]: I see! Great.

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        withBaseCacheDirFor(cacheKey) { cacheDir ->
+            val stateFile = cacheDir.stateFile
+            fileAccessTracker.markAccessed(stateFile)
+            action(stateFile)
+        }
+    }
+
+    class Layout(val fingerprint: File, val state: File)
+
+    fun useForWrite(cacheKey: String, action: (Layout) -> Unit) {
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt:124]: :x: Rename to `useForStore` for symmetry with `useFor*Load`?

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt (hunk 1 above)
Comment (bamboo) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt]: Yes, it does, thank you!

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt (hunk 5 above)
Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt]: Done

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/initialization/InstantExecutionStartParameter.kt [hunk 6]
```diff
@@ -81,7 +82,8 @@ class InstantExecutionStartParameter(
         // and can be used as separators: /, \, :, <, >, ", ?, *, |
         // except we also accept qualified task names with :, so colon is out.
         val cacheKey = StringBuilder()
-        requestedTaskNames.joinTo(cacheKey, separator = "/")
+        cacheKey.append(GradleVersion.current().version)
```
Comment (adammurdoch) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/initialization/InstantExecutionStartParameter.kt:85]: Now that we're using a cross version cache, we should probably use `Hasher` to assemble the hash for the key. Probably should also add an initial cross version test to verify that different Gradle versions do not interfere with each other.

It's also kind of odd that the start parameter knows how to calculate the cache entry key.

**Comment by eskatos** [comment]
Hum, the test introduced in this PR is still flaky, needs more investigation.

**Comment by eskatos** [comment]
@bot-gradle test RFM

**Comment by eskatos** [comment]
@bot-gradle test RFN

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCacheInvalidation.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import org.gradle.instantexecution.initialization.InstantExecutionStartParameter
+import org.gradle.internal.hash.HashValue
+import org.gradle.internal.hash.Hasher
+import org.gradle.internal.hash.Hashing
+import org.gradle.util.GFileUtils
+import org.gradle.util.GradleVersion
+import java.io.File
+
+
+class InstantExecutionCacheInvalidation(
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCacheInvalidation.kt:29]: :x: `InstantExecutionCacheKey` would be a better name.

**Pull request merged**